### DISCUSSION TITLE
Can't get Ctrl+Left/RightArrow to work inside the terminal

### DISCUSSION THREAD

**Discussion by dinvlad** [wezterm/wezterm#4237] [Q&A]
Hi folks, new WezTerm user here, -

I'm using LazyVim, which has the following default mappings for resizing windows in NeoVim:
```lua
map("n", "<C-Left>", "<cmd>vertical resize -2<cr>", { desc = "Decrease window width" })
map("n", "<C-Right>", "<cmd>vertical resize +2<cr>", { desc = "Increase window width" })
```

However, unlike their Ctrl+Up/Down counterparts, they don't appear to work inside WezTerm. When I completely disable the default WezTerm bindings using `config.disable_default_key_bindings = true`, they start working again, which suggests this is indeed an issue with WezTerm defaults, and not something else.

When I view the relevant mappings
```
$ wezterm show-keys --lua | grep -E '(LeftArrow|RightArrow|keys|key_tables|mode)'     
  keys = {
    { key = 'LeftArrow', mods = 'SHIFT|CTRL', action = act.ActivatePaneDirection 'Left' },
    { key = 'LeftArrow', mods = 'SHIFT|ALT|CTRL', action = act.AdjustPaneSize{ 'Left', 1 } },
    { key = 'RightArrow', mods = 'SHIFT|CTRL', action = act.ActivatePaneDirection 'Right' },
    { key = 'RightArrow', mods = 'SHIFT|ALT|CTRL', action = act.AdjustPaneSize{ 'Right', 1 } },
  key_tables = {
    copy_mode = {
      { key = 'LeftArrow', mods = 'NONE', action = act.CopyMode 'MoveLeft' },
      { key = 'LeftArrow', mods = 'ALT', action = act.CopyMode 'MoveBackwardWord' },
      { key = 'RightArrow', mods = 'NONE', action = act.CopyMode 'MoveRight' },
      { key = 'RightArrow', mods = 'ALT', action = act.CopyMode 'MoveForwardWord' },
    search_mode = {
```
they don't seem to include Ctrl+Left/RightArrow explicitly. However, I've tried disabling all of these nevertheless, using
```lua
config.keys = {
	-- Disable defaults
	{ key = "LeftArrow", mods = "SHIFT|CTRL", action = act.DisableDefaultAssignment },
	{ key = "LeftArrow", mods = "SHIFT|ALT|CTRL", action = act.DisableDefaultAssignment },
	{ key = "RightArrow", mods = "SHIFT|CTRL", action = act.DisableDefaultAssignment },
	{ key = "RightArrow", mods = "SHIFT|ALT|CTRL", action = act.DisableDefaultAssignment },
	{ key = "LeftArrow", mods = "NONE", action = act.DisableDefaultAssignment },
	{ key = "LeftArrow", mods = "ALT", action = act.DisableDefaultAssignment },
	{ key = "RightArrow", mods = "NONE", action = act.DisableDefaultAssignment },
	{ key = "RightArrow", mods = "ALT", action = act.DisableDefaultAssignment },
	-- Disable target bindings, just in case
	{ key = "LeftArrow", mods = "CTRL", action = act.DisableDefaultAssignment },
	{ key = "RightArrow", mods = "CTRL", action = act.DisableDefaultAssignment },
}
```
I then see this:
```
$ wezterm show-keys --lua | grep -E '(LeftArrow|RightArrow|keys|key_tables|mode)'
  keys = {
  key_tables = {
    copy_mode = {
      { key = 'LeftArrow', mods = 'NONE', action = act.CopyMode 'MoveLeft' },
      { key = 'LeftArrow', mods = 'ALT', action = act.CopyMode 'MoveBackwardWord' },
      { key = 'RightArrow', mods = 'NONE', action = act.CopyMode 'MoveRight' },
      { key = 'RightArrow', mods = 'ALT', action = act.CopyMode 'MoveForwardWord' },
    search_mode = {
```
so as you can see, most of them got disabled. However, this doesn't seem to help the issue.

Additionally, when I run `wezterm --config debug_key_events=true`, Ctrl+Up/DownArrow events do get logged, but Ctrl+Left/RightArrow don't, as if they're not detected by WezTerm at all.

One other thing I've noticed is WezTerm's menu has `Window -> Select Tab -> Activate the tab to the right` assigned to `^->` by default - even though I cannot find that binding among the default bindings, as you can see above. When I disable it as above though, this assignment disappears, but that doesn't help the issue.

All of this is on macOS Ventura 13.4.1, if that matters.

Thanks in advance for any help!

EDIT: I've looked at other issues prior to posting this, but only now found one that seems relevant: https://github.com/wez/wezterm/issues/3508

**Comment by cstrahan** [comment]
I'm running into precisely the same problem.

Tried these mappings (looked at https://sw.kovidgoyal.net/kitty/keyboard-protocol to figure out the right encoding):

```lua
  { key = 'LeftArrow', mods = 'CTRL', action = act.SendString '\x1b\x5b1;5D' },
  { key = 'RightArrow', mods = 'CTRL', action = act.SendString '\x1b\x5b1;5C' },
```

Used the `kitten` command that comes with the kitty terminal emulator (`kitten show_key -m kitty`) but it looks like CTRL+{LEFT,RIGHT} aren't sent by wezterm, even with the above key maps.

**Comment by cstrahan** [comment]
If you can tolerate using SHFT+CTRL instead of just CTRL (easy enough for me, since I map capslock to control I can roll my pinky over both the shift and control keys), this works:

```lua
  { key = 'LeftArrow', mods = 'SHIFT|CTRL', action = act.SendString '\x1b\x5b1;5D' },
  { key = 'RightArrow', mods = 'SHIFT|CTRL', action = act.SendString '\x1b\x5b1;5C' },
```

But then you'll sacrifice the ability to send SHIFT+CTRL modifiers with the left/right arrow keys. You could make a different sacrifice by using `'SUPER|CTRL'` as the modifiers instead.

Another problem comes to light though, as `kitten show_key -m kitty` shows:

```
ctrl+RIGHT PRESS
CSI 1 ; 5 C

shift+ctrl+RIGHT RELEASE
CSI 1 ; 6 : 3 C
```

Using `SendString` as part of a key mapping will send the correct key press code, but letting go will result in sending the release code for the _actually_ held keys. This could confuse a terminal app that keeps track of held/released key state, though this doesn't apply in the `neovim` case as it is only looking for key presses and ignores the releases.

Perhaps wezterm could use a way to override the key release code as part of an extended `SendString` action, or something along those lines.

**Comment by patcoll** [comment]
> Additionally, when I run `wezterm --config debug_key_events=true`, Ctrl+Up/DownArrow events do get logged, but Ctrl+Left/RightArrow don't, as if they're not detected by WezTerm at all.

I am also running into this same issue. Sonoma 14.5

I am using `config.disable_default_key_bindings = true` to disable all default key bindings because I am still using tmux and trying to slowly test out the waters with wezterm to see how possible it is to completely replace tmux for my workflows.

I'm getting around it by using Karabiner Elements to remap Ctrl+Left/RightArrow to use the Option modifier instead, which I can then re-bind in the wezterm config for my needs.

**Comment by leikoilja** [comment]
I already use `Karabiner` on macos, so i had just added `complex modifications` that helped to pass ctrl+hjkl to the terminal as arrows, essentially making my neovim config work as expected (ctrl+h acting as ctrl+left_arrow to move between windows)
```
{
    "description": "Left ctrl + hjkl to arrow keys Vim (wezterm)",
    "manipulators": [
        {
            "conditions": [
                {
                    "bundle_identifiers": [
                        "^com\\.github\\.wez\\.wezterm$"
                    ],
                    "type": "frontmost_application_unless"
                }
            ],
            "from": {
                "key_code": "h",
                "modifiers": {
                    "mandatory": ["left_control"],
                    "optional": ["any"]
                }
            },
            "to": [{ "key_code": "left_arrow" }],
            "type": "basic"
        },
        {
            "conditions": [
                {
                    "bundle_identifiers": [
                        "^com\\.github\\.wez\\.wezterm$"
                    ],
                    "type": "frontmost_application_unless"
                }
            ],
            "from": {
                "key_code": "j",
                "modifiers": {
                    "mandatory": ["left_control"],
                    "optional": ["any"]
                }
            },
            "to": [{ "key_code": "down_arrow" }],
            "type": "basic"
        },
        {
            "conditions": [
                {
                    "bundle_identifiers": [
                        "^com\\.github\\.wez\\.wezterm$"
                    ],
                    "type": "frontmost_application_unless"
                }
            ],
            "from": {
                "key_code": "k",
                "modifiers": {
                    "mandatory": ["left_control"],
                    "optional": ["any"]
                }
            },
            "to": [{ "key_code": "up_arrow" }],
            "type": "basic"
        },
        {
            "conditions": [
                {
                    "bundle_identifiers": [
                        "^com\\.github\\.wez\\.wezterm$"
                    ],
                    "type": "frontmost_application_unless"
                }
            ],
            "from": {
                "key_code": "l",
                "modifiers": {
                    "mandatory": ["left_control"],
                    "optional": ["any"]
                }
            },
            "to": [{ "key_code": "right_arrow" }],
            "type": "basic"
        }
    ]
}
```

**Comment by sybereal** [comment]
On macOS, by default, Ctrl-Left/Right is globally bound to switching between desktop spaces. If you disable the global shortcut, arrow keys work as normal in WezTerm.

<img width="652" height="423" alt="image" src="https://github.com/user-attachments/assets/db1d7fff-6a77-418f-bd2b-65fd0f8367c8" />
